### PR TITLE
Fix/35134: Auto select category and tag

### PR DESCRIPTION
**Pull request by dukenv0307**
### Details

- Include auto-selection of tags/categories when a single option is available
### Fixed Issues

$ https://github.com/Expensify/App/issues/35134
PROPOSAL: https://github.com/Expensify/App/issues/35134#issuecomment-1909769624

### Tests

Pre-requisite: Violations beta needs to be enabled
1. Set up a workspace with only one tag and one category enabled
2. Open ND. Click on FAB => Choose "Request money" => Choose "Manual"
3. Enter amount
4. Select a workspace that has only one option variable in search category-tag
5. In confirmation screen, verify that the category/tag are auto selected.
- [x] Verify that no errors appear in the JS console

### QA Steps

Pre-requisite: Violations beta needs to be enabled
1. Set up a workspace with only one tag and one category enabled
2. Open ND. Click on FAB => Choose "Request money" => Choose "Manual"
3. Enter amount
4. Select a workspace that has only one option variable in search category-tag
5. In confirmation screen, verify that the category/tag are auto selected.
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[Screencast from 07-02-2024 17:55:51.webm](https://github.com/Expensify/App/assets/129500732/7726e620-caab-4bec-acf4-14768efd592d)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

[Screencast from 06-02-2024 07:50:40.webm](https://github.com/Expensify/App/assets/129500732/9ff4d29d-99d7-4966-b7e6-1e61bb32583e)

</details>

<details>
<summary>iOS: Native</summary>

[Screencast from 06-02-2024 08:20:21.webm](https://github.com/Expensify/App/assets/129500732/aa6a7e59-4589-48ef-905c-310f6d6795f0)

</details>

<details>
<summary>iOS: mWeb Safari</summary>

[Screencast from 06-02-2024 07:53:12.webm](https://github.com/Expensify/App/assets/129500732/41070910-81ad-4305-ae52-6625b3358009)

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

[Screencast from 06-02-2024 07:47:31.webm](https://github.com/Expensify/App/assets/129500732/46a00e20-c623-47c2-bb64-389fdc6516d1)

</details>

<details>
<summary>MacOS: Desktop</summary>

[Screencast from 06-02-2024 08:09:19.webm](https://github.com/Expensify/App/assets/129500732/20af2901-408b-426b-8886-7215432a2343)

</details>

### CODE REVIEW

**Comment by jjcoffee** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.js [hunk 1]
```diff
@@ -477,6 +477,22 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
         IOU.setMoneyRequestMerchant_temporaryForRefactor(transaction.transactionID, distanceMerchant);
     }, [hasRoute, distance, unit, rate, currency, translate, toLocaleDigit, isDistanceRequest, transaction]);
 
+    useEffect(() => {
```
Comment (jjcoffee) [inline src/components/MoneyTemporaryForRefactorRequestConfirmationList.js:480]: We could do with a brief explanatory comment here, I think.

**Comment by jjcoffee** [comment]
@dukenv0307 I think it's a bit clearer if the first test step is "Set up a workspace with only one tag and one category enabled".

**Comment by jjcoffee** [comment]
@dukenv0307 Could you add an Android native screenshot?

**Comment by jjcoffee** [comment]
@dukenv0307 Let's limit this to only auto-select if the category or tag is required (per [this](https://github.com/Expensify/App/issues/35134#issuecomment-1932408100) comment). I've not had feedback on whether or not we should hold for the multi-level tags PR but it might be worth waiting for that anyway before doing any more work here.

**Comment by jjcoffee** [comment]
That [PR](https://github.com/Expensify/App/pull/34983) should be getting merged today.

**Comment by jjcoffee** [comment]
[Merged](https://github.com/Expensify/App/pull/34983#event-11818416613)! @dukenv0307 Are you able to rework this to handle multi-level tags and to only auto-select if the tag is required?

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.js [hunk 2]
```diff
@@ -481,21 +481,29 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
         IOU.setMoneyRequestMerchant(transaction.transactionID, distanceMerchant, true);
     }, [isDistanceRequestWithPendingRoute, hasRoute, distance, unit, rate, currency, translate, toLocaleDigit, isDistanceRequest, transaction]);
 
+    // Auto select the category if there is only one enabled category and it is required
     useEffect(() => {
         const enabledCategories = _.filter(policyCategories, (category) => category.enabled);
-        if (iouCategory || !shouldShowCategories || enabledCategories.length !== 1) {
+        if (iouCategory || !shouldShowCategories || enabledCategories.length !== 1 || !isCategoryRequired) {
```
Comment (jjcoffee) [inline src/components/MoneyTemporaryForRefactorRequestConfirmationList.js:487]: Using `isCategoryRequired` here means we'll get inconsistent behaviour vs tags as `isCategoryRequired` relies on `canUseViolations` being true, whereas the checks for the tags are not reliant on that. Do you think it's better to just use `lodashGet(policy, 'requiresCategory', false)` directly here?

Comment (jjcoffee) [reply src/components/MoneyTemporaryForRefactorRequestConfirmationList.js:487]: @dukenv0307 Ah good point! For multi-level tags I think we need to use the `required` field on each tag. For single-level tags however, we need to use `isTagRequired` as on oldDot this option is only available when multi-level tags are turned off `(People must tag expenses)`:

image

Regarding my previous comment, I think it probably makes sense to just add `canUseViolations` to the tags to keep things consistent. Tags and Categories don't display the `Required` text unless `canUseViolations` is true so it makes sense for us to tie in with that I think!

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.js [hunk 3]
```diff
@@ -493,17 +493,18 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
     // Auto select the tag if there is only one enabled tag and it is required
     useEffect(() => {
         let updatedTagsString = '';
+        const isMultipleLevelsTags = policyTagLists && policyTagLists.length > 1;
         policyTagLists.forEach((tagList, index) => {
             const enabledTags = _.filter(tagList.tags, (tag) => tag.enabled);
-            if (enabledTags.length !== 1 || !tagList.required || TransactionUtils.getTag(transaction, index)) {
+            if (enabledTags.length !== 1 || isMultipleLevelsTags ? !tagList.required : !isTagRequired || TransactionUtils.getTag(transaction, index)) {
```
Comment (jjcoffee) [inline src/components/MoneyTemporaryForRefactorRequestConfirmationList.js:499]:
```suggestion
            if (enabledTags.length !== 1 || !tagList.required || TransactionUtils.getTag(transaction, index)) {
```

Looking at this again, I think `isTagRequired` probably should/will be depreciated anyway as now all tags (including single level) will have the `required` property.

**Comment by jjcoffee** [comment]
Reported an unrelated bug [here](https://expensify.slack.com/archives/C049HHMV9SM/p1708607040166479) where all multi-level tags show as required if the workspace originally had "People must tag expenses" enabled.

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.js (hunk 3 above)
Comment (jjcoffee) [reply src/components/MoneyTemporaryForRefactorRequestConfirmationList.js]: Haha oh sorry I responded on the issue first before seeing this. I'd vote for proceeding with the expectation that single-level tags will also have the `required` field set as the `isTagRequired` field seems obsolete to me. cc @puneetlath in case you have thoughts. (And sorry for all the pings!)

Comment (puneetlath) [reply src/components/MoneyTemporaryForRefactorRequestConfirmationList.js]: Asked here: https://expensify.slack.com/archives/C01GTK53T8Q/p1709227512388819?thread_ts=1708732718.173419&cid=C01GTK53T8Q

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.js [hunk 4]
```diff
@@ -481,6 +482,30 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
...
+        }
+        IOU.setMoneyRequestCategory(transaction.transactionID, enabledCategories[0].name);
+    }, [iouCategory, shouldShowCategories, policyCategories, transaction, isCategoryRequired]);
+
+    // Auto select the tag if there is only one enabled tag and it is required
+    useEffect(() => {
+        let updatedTagsString = TransactionUtils.getTag(transaction);
+        policyTagLists.forEach((tagList, index) => {
+            const enabledTags = _.filter(tagList.tags, (tag) => tag.enabled);
+            if (isUndefined(tagList.required) || enabledTags.length !== 1 || !tagList.required || TransactionUtils.getTag(transaction, index)) {
```
Comment (jjcoffee) [inline src/components/MoneyTemporaryForRefactorRequestConfirmationList.js:499]: How about introducing `isTagListRequired` here, both to keep things tidy but also so that we can use the `canUseViolations` flag for tags too?

**Review by jjcoffee — APPROVED** [review]
LGTM!

**Comment by jjcoffee** [comment]
@dukenv0307 Can you merge main to fix conflicts?

**Comment by jjcoffee** [comment]
@puneetlath To summarise some of the implementation details here:

- Due to the BE issue we've [discussed](https://expensify.slack.com/archives/C01GTK53T8Q/p1709229335363099?thread_ts=1708732718.173419&cid=C01GTK53T8Q) on Slack, auto-select for single-level tags will not work correctly currently. This is because the tag list for them does not have the correct `required` field (it's there, it just gets stuck on either `true` or `false` and never updates). We could use `requiresTag` from the policy for single-level tags as a workaround, but I guess we can expect the BE issue to be fixed at some point!
- Changing the required status of a tag or category on OD whilst a money request screen is open won't revert the auto-selected category/tag. There's no pusher event to support this, so I thought it makes sense to leave it out of the scope of this PR.

**Comment by jjcoffee** [comment]
@dukenv0307 Sorry forgot to say, can you add to the test steps the pre-requisite that the violations beta needs to be enabled? :pray:

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.js [hunk 5]
```diff
@@ -490,6 +491,31 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
...
+        }
+        IOU.setMoneyRequestCategory(transaction.transactionID, enabledCategories[0].name);
+    }, [iouCategory, shouldShowCategories, policyCategories, transaction, isCategoryRequired]);
+
+    // Auto select the tag if there is only one enabled tag and it is required
+    useEffect(() => {
+        let updatedTagsString = TransactionUtils.getTag(transaction);
+        policyTagLists.forEach((tagList, index) => {
+            const enabledTags = _.filter(tagList.tags, (tag) => tag.enabled);
+            const isTagListRequired = isUndefined(tagList.required) ? false : tagList.required && canUseViolations;
```
Comment (puneetlath) [inline src/components/MoneyTemporaryForRefactorRequestConfirmationList.js:508]: What is the `canUseViolations` check for here? Why is this needed?

**Review by puneetlath — CHANGES_REQUESTED** [review]
Just one comment. 

@jjcoffee can you also make sure to report the BE issue in #expensify-bugs so we make sure it gets handled. You can cc me and yuwen.

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.js (hunk 5 above)
Comment (jjcoffee) [reply src/components/MoneyTemporaryForRefactorRequestConfirmationList.js:508]: @puneetlath Tags and categories only display the required label on FE if the violations beta is enabled, so I thought it makes sense to auto-select when it's obvious that the tag/category is required. It's also to keep things consistent with categories which will only autoselect if required + the violations beta is enabled.

**Comment by jjcoffee** [comment]
> can you also make sure to report the BE issue in #expensify-bugs so we make sure it gets handled.
@puneetlath Yuwen has [pushed](https://expensify.slack.com/archives/C01GTK53T8Q/p1709706503772179?thread_ts=1708732718.173419&cid=C01GTK53T8Q) a fix, just waiting for it to hit production.

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.js (hunk 5 above)
Comment (puneetlath) [reply src/components/MoneyTemporaryForRefactorRequestConfirmationList.js:508]: Ok got it.

**Review by puneetlath — APPROVED** [review]
